risc0/zkvm/sdk/rust/guest/src/lib.rs

```diff
@@ -35,34 +34,63 @@ pub mod sha;
 #[cfg(feature = "pure-prove")]
 pub mod sha_insecure;
 
-use core::{arch::asm, mem, panic::PanicInfo, ptr};
+use core::{arch::asm, mem, ptr};
 
 extern "C" {
     fn _fault() -> !;
 }
 
-#[cfg(all(target_arch = "riscv32", not(feature = "std")))]
-#[panic_handler]
-unsafe fn panic_fault(panic_info: &PanicInfo<'static>) -> ! {
+/// Aborts the guest with the given message.  This message should
+/// include a trailing \0.
+pub unsafe fn _fail(msg: &str) -> ! {
     use risc0_zkvm_platform::io::GPIO_FAULT;
-
-    let msg = alloc_crate::format!("{}\0", panic_info);
     let ptr = msg.as_ptr();
     memory_barrier(ptr);
+
     // A compliant host should fault when it receives this descriptor.
-    GPIO_FAULT.as_ptr().write_volatile(ptr);
+    unsafe { GPIO_FAULT.as_ptr().write_volatile(ptr) }
 
     // As a fallback for uncompliant hosts, force an unaligned write, which causes a
     // fault within the Risc0 VM.
-    _fault()
+    unsafe {
+        _fault();
+    }
+
+    unreachable!();
+}
+
+/// This defines standalone (no_std) panic and alloc error handlers in
+/// a guest method, for guest methods that do not want to use the
+/// standard library.
+///
+/// Guest methods will also need to specify these at the beginning of the file:
+///
+/// #![no_std]
+/// #![feature(alloc_error_handler)]
```

Comment:
Won't this require guests to always use a nightly compiler even after we get the new target triple added to upstream?

Reply:
Only for no_std though